v2/tools/generator/internal/codegen/pipeline/add_configmaps.go

```diff
@@ -0,0 +1,152 @@
+/*
+ * Copyright (c) Microsoft Corporation.
+ * Licensed under the MIT license.
+ */
+
+package pipeline
+
+import (
+	"context"
+
+	"github.com/pkg/errors"
+
+	"github.com/Azure/azure-service-operator/v2/tools/generator/internal/astmodel"
+	"github.com/Azure/azure-service-operator/v2/tools/generator/internal/config"
+)
+
+// AddConfigMapsStageID is the unique identifier for this pipeline stage
+const AddConfigMapsStageID = "addConfigMaps"
+
+// AddConfigMaps replaces properties flagged as a config map with genruntime.ConfigMapReference or genruntime.OptionalConfigMapReference
+func AddConfigMaps(config *config.Configuration) *Stage {
+	stage := NewStage(
+		AddConfigMapsStageID,
+		"Replace properties flagged as a configMap with genruntime.ConfigMapReference, and properties "+
+			"flagged as an optional configMap with genruntime.OptionalConfigMapReference",
+		func(ctx context.Context, state *State) (*State, error) {
+			defs := state.Definitions()
+
+			updatedDefs, err := transformConfigMaps(config, defs)
+			if err != nil {
+				return nil, errors.Wrap(err, "transforming spec secrets")
+			}
+
+			return state.WithDefinitions(updatedDefs), nil
+		})
+
+	stage.RequiresPostrequisiteStages(CreateARMTypesStageID)
+
+	return stage
+}
+
+func transformPropertyToConfigMapReference(prop *astmodel.PropertyDefinition, newType astmodel.Type) (*astmodel.PropertyDefinition, error) {
+	// The expectation is that this is a string
+	propType := prop.PropertyType()
+	if !astmodel.Unwrap(propType).Equals(astmodel.StringType, astmodel.EqualityOverrides{}) {
+		return nil, errors.Errorf("expected property %q to be a string, but was: %T", prop.PropertyName(), propType)
+	}
+
+	// check if it's optional
+	_, optional := astmodel.AsOptionalType(propType)
+
+	if optional {
+		newType = astmodel.NewOptionalType(newType)
+	}
+
+	return prop.WithType(newType), nil
+}
+
+func createNewConfigMapReference(prop *astmodel.PropertyDefinition, newType astmodel.Type) (*astmodel.PropertyDefinition, *astmodel.PropertyDefinition, error) {
+	// The expectation is that this is a string
+	propType := prop.PropertyType()
+	if !astmodel.Unwrap(propType).Equals(astmodel.StringType, astmodel.EqualityOverrides{}) {
+		return nil, nil, errors.Errorf("expected property %q to be a string, but was: %T", prop.PropertyName(), propType)
+	}
+
+	// check if it's optional
+	_, optional := astmodel.AsOptionalType(propType)
+
+	if optional {
+		newType = astmodel.NewOptionalType(newType)
+	}
+
```

Comment:
Good catch. Not sure why our linting didn't notice that. I actually don't need this check at all anymore since I have to force them all to be optional no matter what. (Because even an originally required property will be optional now since it's `oneof{A, AConfigRef}` pretty much.

I've removed these statements.